### PR TITLE
Allow disabling process killing on interruption

### PR DESCRIPTION
**Pull request by apetres**
Plugins may want to do custom action when the execution of a remote process is interrupted. Currently it is not possible, because
the process started on a slave is killed on interruption.
My actual use case: https://groups.google.com/forum/#!msg/jenkinsci-dev/s6IkynOzdqE/2eFNKN9lAAAJ

My solution is to have a flag called killWhenInterrupted, set by default to true, which prevents process killing on the slave when unset.

### CODE REVIEW

File: core/src/main/java/hudson/Launcher.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Launcher.java]: Needs to be removed before merge

File: core/src/main/java/hudson/Launcher.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Launcher.java]: LGTM, no external usages in detached components and plugins.

File: core/src/main/java/hudson/Proc.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Proc.java]: I would suggest using `final private` fields and getters

File: core/src/main/java/hudson/Proc.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Proc.java]: Would be preferable to just call the new constructor below to avoid the logic duplication

File: core/src/main/java/hudson/Launcher.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Launcher.java]: 🐛 it causes a regression in external API users. The default value will be `false` though the logic was always killing processes before the change

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
`Launcher` should retain the default behavior when `killWhenInterrupted` is not invoked. My suggestion would be to invert flags and to use `dontKillWhenInterrupted` or so

File: core/src/main/java/hudson/Launcher.java
Comment (apetres) [reply core/src/main/java/hudson/Launcher.java]: You are right. Flag was inverted.

File: core/src/main/java/hudson/Launcher.java
Comment (apetres) [reply core/src/main/java/hudson/Launcher.java]: Done

File: core/src/main/java/hudson/Proc.java
Comment (apetres) [reply core/src/main/java/hudson/Proc.java]: It can't be `final` since the value is set in the implementations.
Also it can't be `private` in order to be accessible from implementations. Maybe I should add a setter and make it private?

File: core/src/main/java/hudson/Proc.java
Comment (apetres) [reply core/src/main/java/hudson/Proc.java]: Done

**Review by apetres — COMMENTED** [review]
killWhenInterrupted flag was inverted.

File: core/src/main/java/hudson/Launcher.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Launcher.java]: It would be preferable to keep refactorings in a separate pull request. But this PR is a non-backportable feature, so I can live with that

File: core/src/main/java/hudson/Launcher.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Launcher.java]: Add some Javadoc with `@since TODO`

File: core/src/main/java/hudson/Proc.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Proc.java]: `@since TODO`. Protected fields are a part of API

File: core/src/main/java/hudson/Launcher.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Launcher.java]: Some documentation would be useful + `@since TODO`

**Review by oleg-nenashev — COMMENTED** [review]
The code looks good to me. There are more formatting cleanup than real code though.

@apetres BTW, have you already prototyped a patch for your plugin with this change? If yes, it would be nice to have a link

**Comment by oleg-nenashev** [comment]
Happy to approve once Javadoc is added

**Comment by daniel-beck** [comment]
How does this compare to `ProcessKillingVeto`?

File: core/src/main/java/hudson/Launcher.java
Comment (apetres) [reply core/src/main/java/hudson/Launcher.java]: Sorry for that, my IDE is on steroids. I reverted the refactorings.

File: core/src/main/java/hudson/Launcher.java
Comment (apetres) [reply core/src/main/java/hudson/Launcher.java]: Done

File: core/src/main/java/hudson/Launcher.java
Comment (apetres) [reply core/src/main/java/hudson/Launcher.java]: Done

File: core/src/main/java/hudson/Proc.java
Comment (apetres) [reply core/src/main/java/hudson/Proc.java]: Done

**Review by apetres — COMMENTED** [review]
@oleg-nenashev here is the link to the prototype plugin: https://github.com/apetres/testcomplete-plugin/commit/566ce25f22d89088d8ac06c2ee2e32d3406f16a7

**Comment by apetres** [comment]
@daniel-beck AFAICT `ProcessKillingVeto` always prevents the killing of a process with given pid/arguments,etc. This is an undesired behavior for me because I do want to kill the process, but only after I executed custom action on the caller (plugin) side, see the link for the prototype implementation above.

**Review by oleg-nenashev — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
Thanks @apetres !

**Comment by daniel-beck** [comment]
No proposed changelog entry 👎

**Comment by daniel-beck** [comment]
> always prevents the killing of a process with given pid/arguments,etc. This is an undesired behavior for me because I do want to kill the process, but only after I executed custom action on the caller (plugin) side, see the link for the prototype implementation above.

I don't see the problem with that `BUILD_ID` is among the env vars of the `IOSProcess` you'd veto, so you can trivially detect "your" process -- and, while vetoing the regular killing, implement your own as you did.

IMO this PR is not well justified was merged prematurely.

**Comment by oleg-nenashev** [comment]
@daniel-beck Actually checking environment of a standalone process is a big problem for Windows. In Jenkins this resolution *most probably* does not work properly for 32-bit systems now: https://github.com/kohsuke/winp/issues/48

So I would not rely on this logic in Jenkins now.

But I agree that I merged it too prematurely, sorry about that

**Comment by daniel-beck** [comment]
@oleg-nenashev No worries, easy to revert and seems straightforward to submit again for further review.

**Comment by daniel-beck** [comment]
I'd also really like to see a test for this feature in core, if reasonably possible.

**Comment by daniel-beck** [comment]
I merged the revert PR https://github.com/jenkinsci/jenkins/pull/3339

@apetres This proposed change is **not** rejected, it was just prematurely merged, and given our release frequency and commitment to maintain API compatibility, not reverting would equal permanent acceptance starting later today with the release of 2.111. I'm sorry about this mess.

We're of course open to a resubmission of this PR against the current master, ideally with the concerns expressed above addressed.

**Comment by apetres** [comment]
Ok, so I should:
- propose a changelog entry
- add a test in core

@daniel-beck about your concerns above:

> you can trivially detect "your" process

Yes, I can. Since only one instance of TestComplete/TestExecute process can be run at a time, I can even detect it using the arguments, something like [here](https://github.com/jenkinsci/msbuild-plugin/blob/master/src/main/java/hudson/plugins/msbuild/MsBuildKillingVeto.java). So I don't have to rely on env vars.

> while vetoing the regular killing, implement your own as you did

It's not clear for me how to kill a vetoed process from the plugin's side, since it only has access to the _Proc_ object. Also IMO _killing a vetoed process_ contradicts the concept of vetoing. Maybe I could run native killing instructions in a separate, joined process, after I ran my custom code (but IMO that's not really nice).

**Comment by daniel-beck** [comment]
@apetres Tests would be great! In general, what we expect from a PR is noted in the PR template at https://github.com/jenkinsci/jenkins/blob/master/.github/PULL_REQUEST_TEMPLATE.md

I think a reasonable start to continue this review would be the existing code + tests. Note that you'll probably have to re-do the code changes since it was merged and reverted (unless you're a Git wizard and know things I don't).

Additionally, an explanation why `ProcessKillingVeto` would not work for your use case would be good. I don't see how using that to *not* just kill the process and implement your own logic is inferior to this approach (although @oleg-nenashev makes a good argument there).

**Comment by apetres** [comment]
@daniel-beck `ProcessKillingVeto` is used to prevent killing processes. That means that they keep running even after the build is finished. From my plugin's side there is no way to kill the vetoed process because `Proc.kill()` won't do anything because of the veto.

For me this is an undesired side effect because I _do_ want to kill the process after I run some logic on the plugin's side. Keeping it running after the build would cause an issue with future builds running on the same executor.

This is how the logic my plugin implementation looks like, for more details check the link to the prototype implementation.

```
Proc process;
try {
    Launcher.ProcStarter processStarter = launcher.launch()
        .cmds(args)
    //  .dontKillWhenInterrupted(true);
    process = processStarter.start();
    process.join();
    processTestResults()
} catch (InterruptedException e) {
   // InterruptedException usually is caught earlier on the executor's side, resulting in
   // TestComplete being killed and not saving the rest results correctly. That way I end
   // up without any test results, having ran hours of tests for nothing. That is the reason
   // I don't want to kill it
   stopTestCompleteNormally();
   processTestResults();
   // if I implement ProcessKillingVeto for the TestComplete process the line below won't
   // have any effect due to the veto. The veto prevent the killing of the process
   // on the executor side too which may result in a rogue TestComplete if stopTestCompleteNormally()
   // fails killing the process.
   process.kill();
   //Let Jenkins mark it as ABORTED
   throw e;
}
```

Now that I take a second look maybe implementing `ProcessKillingVeto` plus some custom native killing instructions (instead of `process.kill()`) would do the job. But IMO it would be much more nice with a flag on `Proc`'s side.

**Comment by daniel-beck** [comment]
Or just add a flag that a given process should be killable after being processed by your custom logic.

Still, this approach does seem nicer, so I don't consider that a blocker.

**Comment by apetres** [comment]
Created a new PR with the desired changes.

**Pull request merged**
